include/nanobind/stl/chrono.h

```diff
@@ -0,0 +1,439 @@
+        }
+        // Leak references to these datetime types. We could improve upon
+        // this by storing them in the internals structure and decref'ing
+        // in internals_cleanup(), but it doesn't seem worthwhile for
+        // something this fundamental. We can't store nb::object in this
+        // structure because it might be destroyed after the Python
+        // interpreter has finalized.
+        datetime = datetime_o.release();
+        time = time_o.release();
+        date = date_o.release();
+        timedelta = timedelta_o.release();
+        return true;
+    }
+};
+
+inline datetime_types_t datetime_types;
+
+// The next few functions are marked 'inline' for linkage purposes
+// (since they might be in multiple translation units and the linker
+// should pick one) but NB_NOINLINE because we don't want the bloat
+// of actually inlining them. They are defined in this header instead
+// of in the built nanobind library in order to avoid increasing the
+// library size for users who don't care about datetimes.
+
+NB_NOINLINE inline bool set_from_int_attr(int *dest, PyObject *o,
+                                          const char *name) {
+    PyObject *value = PyObject_GetAttrString(o, name);
+    if (!value) return false;
+    long lval = PyLong_AsLong(value);
+    if (lval == -1 && PyErr_Occurred()) {
+        Py_DECREF(value);
+        return false;
+    }
+    Py_CLEAR(value);
+    if (lval < std::numeric_limits<int>::min() ||
+        lval > std::numeric_limits<int>::max()) {
+        PyErr_Format(PyExc_OverflowError,
+                     "%R attribute '%s' (%R) does not fit in an int",
+                     o, name, value);
+        return false;
+    }
+    *dest = static_cast<int>(lval);
+    return true;
+}
+
+#endif // Py_LIMITED_API
+
+// Unpack a datetime.timedelta object into integer days, seconds, and
+// microseconds. Returns true if successful, false if `o` is not a timedelta,
+// or false and sets the Python error indicator if something else went wrong.
+NB_NOINLINE inline bool unpack_timedelta(PyObject *o, int *days,
+                                         int *secs, int *usecs) noexcept {
+#ifndef Py_LIMITED_API
+    if (!PyDateTimeAPI) {
+        PyDateTime_IMPORT;
+        if (!PyDateTimeAPI) {
+            return false;
+        }
+    }
+    if (PyDelta_Check(o)) {
+        *days = PyDateTime_DELTA_GET_DAYS(o);
+        *secs = PyDateTime_DELTA_GET_SECONDS(o);
+        *usecs = PyDateTime_DELTA_GET_MICROSECONDS(o);
+        return true;
+    }
+#else
+    if (!datetime_types && !datetime_types.do_import()) {
+        return false;
+    }
+    if (int is_td = PyObject_IsInstance(o, datetime_types.timedelta.ptr());
+        is_td < 0) {
+        return false;
+    } else if (is_td > 0) {
+        return (set_from_int_attr(days, o, "days") &&
+                set_from_int_attr(secs, o, "seconds") &&
+                set_from_int_attr(usecs, o, "microseconds"));
+    }
+#endif
+    return false;
```

Comment:
One general observation about this function is that some failure return values have a Python error state set, and others don't. Not sure if that's a problem.

Reply:
This is intentional and is mentioned in the comments; I wanted to distinguish between "not the right type" errors (which are common during overload resolution and shouldn't produce a warning) and "something unexpected occurred" errors (which should be user-visible to reduce confusion if possible). However I noticed I was returning from `type_caster::from_python()` with an error set, which I think is unkosher; I've changed that to use `PyErr_WriteUnraisable`.